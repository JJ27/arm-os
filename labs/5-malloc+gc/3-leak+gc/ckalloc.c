// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.
#include "rpi.h"
#include "ckalloc.h"
#include "kr-malloc.h"

unsigned ck_verbose_p = 0;

// should hold all allocated blocks
static hdr_t *alloc_list; 

// returns pointer to the first allocated header block.
hdr_t *ck_first_alloc(void) {
    return alloc_list;
}

// return header associated with <ptr> if one exists.
hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_alloc(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) 
            return h;
    return 0;
}


/***********************************************************************
 * implement the rest
 */

// is <ptr> inside <h>'s data block?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    if(h->state != ALLOCED) {
        panic("should only have allocated blocks: sdtate=%d\n", h->state);
        return 0;
    }

    // use ck_data_start/_end 
    todo("check that <ptr> is in data for <h>\n");
}


// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = ck_ptr_is_alloced(addr);
    if(!h)
        loc_panic(l, "freeing bogus pointer: %p\n", addr);

    // allocated block starts right after the header.

    void *blk_start = ck_data_start(h);
    if(blk_start != addr)
        loc_panic(l, "not freeing using start pointer: have %p, need %p\n",
            addr, blk_start);

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    if(ck_verbose_p)
        loc_debug(l, "freeing %p\n", addr);

    assert(ck_ptr_is_alloced(addr));
    h->state = FREED;

    // just remove from the allocated list.
    todo("implement the rest\n");

    kr_free(h);
}


// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t l) {
    static unsigned block_id=1;

    hdr_t *h = kr_malloc(nbytes + sizeof *h);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = l;
    h->block_id = block_id++;

    // set addr;
    void *addr = 0;

    todo("put on allocated list\n");

    assert(ck_ptr_is_alloced(addr));
    if(ck_verbose_p)
        loc_debug(l, "successful alloc of %p\n", addr);
    return addr;
}
