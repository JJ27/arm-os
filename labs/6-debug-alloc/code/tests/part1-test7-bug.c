// a bit more complicated:
//  1. do N random size allocations.
//  2. make sure there aren't errors.
//  3. do N random legal writes
//  4. make sure there aren't errors.
//  5. free everything
//  6. make sure there aren't errors.
//  7. do random illegal writes, record in <count>
//  8. make sure there are <count> errors.

// use after free.
#include "rpi.h"
#include "ckalloc.h"

void notmain(void) {
    printk("test4: use after free\n");

    ck_verbose_set(0);
    enum { N  = 64 };

    //  1. do N random size allocations.
    void **vec = ckalloc(sizeof *vec * N);
    for(int i = 0; i < N; i++)
        vec[i] = ckalloc(N);

    //  2. make sure there aren't errors.
    if(ck_heap_errors())
        panic("invalid error!!\n");
    else
        trace("SUCCESS heap checked out\n");

    //  3. do N random legal writes
    for(int i = 0; i < N; i++) {
        put32(vec[i] + rpi_rand32() % i, i);
        //  4. make sure there aren't errors.
        if(ck_heap_errors())
            panic("invalid error!!\n");
    }

    //  5. free everything
    for(int i = 0; i < N; i++)
        ckfree(vec[i]);

    // heh.  can't do this one b/c will kill
    // the pointers.
    // ckfree(vec);

    //  6. make sure there aren't errors.
    if(ck_heap_errors())
        panic("invalid error!!\n");

    //  7. do random illegal writes, record in <count>
    unsigned exp_errors = 0;
    for(int i = 0; i < N; i++) {
        unsigned rz1_bug =  rpi_rand32() % REDZONE_NBYTES;
        unsigned rz2_bug =  rpi_rand32() % REDZONE_NBYTES;

        if(rpi_rand32() % 2) {
            if(rz1_bug) {
                put8(vec[i] - rz1_bug, 0);
                exp_errors++;
            }
        } else {
            if(rz2_bug) {
                put8(vec[i] + i + rz2_bug, 0);
                exp_errors++;
            }
        }
    }

    //  8. make sure there are <count> errors.
    int n = ck_heap_errors();
    if(n != exp_errors)
        panic("invalid number of errors, expected %d, nave %d\n", n, N*2);

    trace("SUCCESS heap checked out\n");
}
