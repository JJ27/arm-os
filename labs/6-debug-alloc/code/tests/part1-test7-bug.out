TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:notmain:SUCCESS heap checked out
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:ck_heap_errors:SUCCESS: checked 65 blocks, detected no errors
TRACE:ck_heap_errors:going to check heap
TRACE:mem_check:ERROR:Freed block 65  corrupted at offset -62
TRACE:hdr_print:	logical block id=65,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 64  corrupted at offset 159
TRACE:hdr_print:	logical block id=64,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 63  corrupted at offset -86
TRACE:hdr_print:	logical block id=63,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 62  corrupted at offset 157
TRACE:hdr_print:	logical block id=62,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 61  corrupted at offset 149
TRACE:hdr_print:	logical block id=61,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 60  corrupted at offset -85
TRACE:hdr_print:	logical block id=60,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 59  corrupted at offset 138
TRACE:hdr_print:	logical block id=59,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 58  corrupted at offset -81
TRACE:hdr_print:	logical block id=58,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 57  corrupted at offset 79
TRACE:hdr_print:	logical block id=57,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 56  corrupted at offset -112
TRACE:hdr_print:	logical block id=56,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 55  corrupted at offset 63
TRACE:hdr_print:	logical block id=55,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 54  corrupted at offset -26
TRACE:hdr_print:	logical block id=54,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 53  corrupted at offset 84
TRACE:hdr_print:	logical block id=53,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 52  corrupted at offset 172
TRACE:hdr_print:	logical block id=52,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 51  corrupted at offset 76
TRACE:hdr_print:	logical block id=51,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 50  corrupted at offset -118
TRACE:hdr_print:	logical block id=50,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 49  corrupted at offset -30
TRACE:hdr_print:	logical block id=49,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 48  corrupted at offset -31
TRACE:hdr_print:	logical block id=48,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 47  corrupted at offset 165
TRACE:hdr_print:	logical block id=47,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 46  corrupted at offset 79
TRACE:hdr_print:	logical block id=46,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 45  corrupted at offset 164
TRACE:hdr_print:	logical block id=45,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 44  corrupted at offset 117
TRACE:hdr_print:	logical block id=44,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 43  corrupted at offset 162
TRACE:hdr_print:	logical block id=43,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 42  corrupted at offset -49
TRACE:hdr_print:	logical block id=42,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 41  corrupted at offset 56
TRACE:hdr_print:	logical block id=41,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 40  corrupted at offset 129
TRACE:hdr_print:	logical block id=40,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 39  corrupted at offset 135
TRACE:hdr_print:	logical block id=39,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 38  corrupted at offset 93
TRACE:hdr_print:	logical block id=38,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 37  corrupted at offset -23
TRACE:hdr_print:	logical block id=37,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 36  corrupted at offset 148
TRACE:hdr_print:	logical block id=36,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 35  corrupted at offset 89
TRACE:hdr_print:	logical block id=35,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 34  corrupted at offset 73
TRACE:hdr_print:	logical block id=34,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 33  corrupted at offset 96
TRACE:hdr_print:	logical block id=33,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 32  corrupted at offset 145
TRACE:hdr_print:	logical block id=32,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 31  corrupted at offset -25
TRACE:hdr_print:	logical block id=31,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 30  corrupted at offset -127
TRACE:hdr_print:	logical block id=30,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 29  corrupted at offset 142
TRACE:hdr_print:	logical block id=29,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 28  corrupted at offset 107
TRACE:hdr_print:	logical block id=28,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 27  corrupted at offset -18
TRACE:hdr_print:	logical block id=27,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 26  corrupted at offset 67
TRACE:hdr_print:	logical block id=26,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 25  corrupted at offset 136
TRACE:hdr_print:	logical block id=25,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 24  corrupted at offset -16
TRACE:hdr_print:	logical block id=24,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 23  corrupted at offset -25
TRACE:hdr_print:	logical block id=23,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 22  corrupted at offset -123
TRACE:hdr_print:	logical block id=22,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 21  corrupted at offset -90
TRACE:hdr_print:	logical block id=21,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 20  corrupted at offset 61
TRACE:hdr_print:	logical block id=20,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 19  corrupted at offset -19
TRACE:hdr_print:	logical block id=19,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 18  corrupted at offset 74
TRACE:hdr_print:	logical block id=18,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 17  corrupted at offset -19
TRACE:hdr_print:	logical block id=17,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 16  corrupted at offset 62
TRACE:hdr_print:	logical block id=16,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 15  corrupted at offset 22
TRACE:hdr_print:	logical block id=15,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 14  corrupted at offset -116
TRACE:hdr_print:	logical block id=14,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 13  corrupted at offset 77
TRACE:hdr_print:	logical block id=13,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 12  corrupted at offset -88
TRACE:hdr_print:	logical block id=12,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 11  corrupted at offset 20
TRACE:hdr_print:	logical block id=11,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 10  corrupted at offset -123
TRACE:hdr_print:	logical block id=10,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 9  corrupted at offset -118
TRACE:hdr_print:	logical block id=9,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 8  corrupted at offset 102
TRACE:hdr_print:	logical block id=8,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 7  corrupted at offset 53
TRACE:hdr_print:	logical block id=7,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 6  corrupted at offset 20
TRACE:hdr_print:	logical block id=6,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 5  corrupted at offset -51
TRACE:hdr_print:	logical block id=5,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:mem_check:ERROR:Freed block 4  corrupted at offset 52
TRACE:hdr_print:	logical block id=4,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 3  corrupted at offset 17
TRACE:hdr_print:	logical block id=3,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:check_list:	Wrote block after free!
TRACE:mem_check:ERROR:Freed block 2  corrupted at offset -80
TRACE:hdr_print:	logical block id=2,  nbytes=64
TRACE:hdr_print:	Block allocated at: tests/part1-test7-bug.c:notmain:24
TRACE:ck_heap_errors:checked 65 blocks, detected 64 errors
TRACE:notmain:SUCCESS heap checked out
